packages/cicero-core/src/contractinstance.js

```diff
@@ -62,8 +65,142 @@ class ContractInstance extends Instance {
      * @param {object} options - additional options
      * @return {Promise<ContractInstance>} a Promise to the instance
      */
-    static fromArchive(buffer, options) {
-        return InstanceLoader.fromArchive(ContractInstance, buffer, options);
+    static async fromArchive(buffer, options) {
+        const instance = await InstanceLoader.fromArchive(ContractInstance, buffer, options);
+        return instance;
+    }
+
+    /**
+     * Sign a smart legal contract.
+     * @param {String} p12File - encoded string of p12 keystore file
+     * @param {String} passphrase - passphrase for the keystore file
+     * @param {String} signatory - name of the signatory
+     * @return {Promise<Buffer>} the zlib buffer
+     */
+    async signContract(p12File, passphrase, signatory) {
+        if (this.contractSignatures.length !== 0 ) {
+            this.verifySignatures();
+        }
+        const timestamp = Date.now();
+        this.sign(p12File, passphrase, timestamp, signatory);
+        return await this.toArchive('ergo');
+    }
+
+    /**
+     * signs a string made up of contract hash and time stamp using private key derived
+     * from the keystore
+     * @param {String} p12File - encoded string of p12 keystore file
+     * @param {String} passphrase - passphrase for the keystore file
+     * @param {Number} timestamp - timestamp of the moment of signature is done
+     * @param {String} signatory - name of the signatory
+     * @ppublic
+     */
+    sign(p12File, passphrase, timestamp, signatory) {
+        if (typeof(p12File) !== 'string') {throw new Error('p12File should be of type String!');}
+        if (typeof(passphrase) !== 'string') {throw new Error('passphrase should be of type String!');}
+        if (typeof(timestamp) !== 'number') {throw new Error('timeStamp should be of type Number!');}
+
+        const instanceHash = this.getHash();
+        // decode p12 from base64
+        const p12Der = forge.util.decode64(p12File);
+        // get p12 as ASN.1 object
+        const p12Asn1 = forge.asn1.fromDer(p12Der);
+        // decrypt p12 using the passphrase 'password'
+        const p12 = forge.pkcs12.pkcs12FromAsn1(p12Asn1, false, passphrase);
+        //X509 cert forge type
+        const certificateForge = p12.safeContents[0].safeBags[0].cert;
```

Comment:
Under what circumstances would we ever need to access entries that are not in the first position of `safeContents` or `safeBags` array?

Reply:
I don't think we'll ever have to, it's mostly metadata of the keystore stored in it.

Reply:
How safe is this path really? What happens if there is no `safeContent` or no `safeBags`. Could you clarify what the forge API supports there?

Reply:
@mttrbrts `safeContents` contains all the things present inside the keystore. In this case at the first position it is the certificate and in the second position it is the private key. So `safeContents[0].safeBags[0]` contains the x509 cert, and only has one object inside the safeBags array. Similary `safeContents[1].safeBags[0]` has the private key and also has only object inside the safeBags array. @jeromesimeon there is another method `getBags()` using which the the material inside safe bags can be accessed but either one of the following property is to be known:

- friendlyName
- localKeyIdHex
- bagType

Reply:
- I was not able to find a friendly name
- `{ localKeyId: [ '\u0010÷&Þ\u001e°ÑÆ\u0015Ã?\u0011Aw\u0010¥' ] }`, this was the localKeyId
-  `1.2.840.113549.1.12.10.1.3`, this was the bag type for a x509 cert & `1.2.840.113549.1.12.10.1.2` this for a private key.

Since these were a bit hard to understand I decided to take the route of selecting the certificate and the private key from `safeContents` and `safeBags` directly

Reply:
You can refer to this replit: https://replit.com/@sanketshevkar/forgeBags#index.js